Supporting spaced names and empty payload

ssouzawallace opened the pull request:
### **The problem**
Many analytics tools provide some flexibility naming events and attributes. With this in mind, this PR is intended to add this feature to JustTracker.

### **Bonus**
Additionally, I've removed all `for index in 0...keys.count-1` and some forced unwraps in order to improve code safety. Now, if the user doesn't provide a payload (what isn't so uncommon), the generator will do the magic without break 😃

### **Result**
For the Events.plist in the end of this description we have this generated class:
```swift
@objc public class JEEventMyCoolEvent:NSObject,JEEvent {

    //JEEvent protocol
    public let name: String = "My Cool Event"

    public var payload: Payload {
        return [kCoolerSpacedAttribute: coolerSpacedAttribute as NSObject, kAnotherAttribute: anotherAttribute as NSObject, kAttribute: attribute as NSObject, kGiveMeSpaAAACe: giveMeSpaAAACe as NSObject]
    }

    public var registeredTrackers: [String] {
        return ["console", "tracker2"]
    }

    //keys
    private let kCoolerSpacedAttribute = "Cooler Spaced Attribute"
    private let kAnotherAttribute = "anotherAttribute"
    private let kAttribute = "Attribute"
    private let kGiveMeSpaAAACe = "give me spa a  a   a    ce"

    var coolerSpacedAttribute: String = ""
    var anotherAttribute: String = ""
    var attribute: String = ""
    var giveMeSpaAAACe: String = ""

    public init(coolerSpacedAttribute: String, anotherAttribute: String, attribute: String, giveMeSpaAAACe: String) {
        super.init()
        self.coolerSpacedAttribute = coolerSpacedAttribute
        self.anotherAttribute = anotherAttribute
        self.attribute = attribute
        self.giveMeSpaAAACe = giveMeSpaAAACe
    }
}
```
Without this PR, the result is:

```swift
@objc public class JEEventMy Cool Event:NSObject,JEEvent {

    //JEEvent protocol
    public let name: String = "My Cool Event"

    public var payload: Payload {
        return [kCooler Spaced Attribute : Cooler Spaced Attribute as NSObject, kAnotherAttribute : anotherAttribute as NSObject, kAttribute : Attribute as NSObject, kGive me spa a  a   a    ce : give me spa a  a   a    ce as NSObject]
    }

    public var registeredTrackers: [String] {
        return ["console", "tracker2"]
    }

    //keys
    private let kCooler Spaced Attribute = "Cooler Spaced Attribute"
    private let kAnotherAttribute = "anotherAttribute"
    private let kAttribute = "Attribute"
    private let kGive me spa a  a   a    ce = "give me spa a  a   a    ce"

    var Cooler Spaced Attribute: String = ""
    var anotherAttribute: String = ""
    var Attribute: String = ""
    var give me spa a  a   a    ce: String = ""

    public init(Cooler Spaced Attribute: String, anotherAttribute: String, Attribute: String, give me spa a  a   a    ce: String) {
        super.init()
        self.Cooler Spaced Attribute = Cooler Spaced Attribute
        self.anotherAttribute = anotherAttribute
        self.Attribute = Attribute
        self.give me spa a  a   a    ce = give me spa a  a   a    ce
    }
}
```

Events.plist:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>events</key>
	<dict>
		<key>My Cool Event</key>
		<dict>
			<key>registeredTrackers</key>
			<array>
				<string>console</string>
				<string>tracker2</string>
			</array>
			<key>payloadKeys</key>
			<array>
				<string>Cooler Spaced Attribute</string>
				<string>anotherAttribute</string>
				<string>Attribute</string>
				<string>give me spa a  a   a    ce</string>
			</array>
		</dict>		
	</dict>
</dict>
</plist>
```

ssouzawallace committed:
Enabling names with space

Adding support for spaced names in events and attributes.

Also, removing some forced unwraps and index based iterations that was preventing use of events without payload. What could be one use case.

ssouzawallace committed:
Fixing payload array generation

federicocappelli commented on JustTrack/JEEventsGenerator/main.swift:
```diff
@@ -9,26 +9,26 @@
-    case eventInit = "JEEventInitTemplate.jet"
-    case eventInitAssignsList = "JEEventInitAssignTemplate.jet"
-    case eventInitParam = "JEEventInitParam.jet"
+    case eventList              = "JEEventListTemplate.jet"
```
This code style doesn't follow the Just Eat, and the official swift, guidelines, the prefered spacing style is:
`case eventList = "JEEventListTemplate.jet"`

ssouzawallace replied:
I have a bit of OCD 😂

Will fix later.

federicocappelli commented on JustTrack/JEEventsGenerator/main.swift:
```diff
@@ -153,70 +153,77 @@ func generateStructs(_ events: [String : AnyObject]) throws -> NSString {
-        /*
-         <*event_init*> =
-         init(<*event_init_params*>) {
+        if let eventDic = events[eventName] as? [String : AnyObject] {
```
if the eventName exist then the dictionary must exist, even if empty.
Could you please remove the if and revert to the previous implementation?
This will even make easier to review the PR, highlighting only the actual changes and not the entire method.

ssouzawallace replied:
I agree that event names and trackers should be mandatory. Will fix later.

But in this case the error should be clear to the user. Clearer than debugging the .plist with hundreds of events, as mentioned in the @federicocappelli comment.

federicocappelli commented on JustTrack/JEEventsGenerator/main.swift:
```diff
@@ -153,70 +153,77 @@ func generateStructs(_ events: [String : AnyObject]) throws -> NSString {
+             self.test2 = test2
+             self.test3 = test3
+             */
+            var eventInit: String = try generateEventInit(keys)
```
The code generated is pretty ugly in case of empty events:

```
@objc public class JEEventEmptyEvent:NSObject,JEEvent {

    //JEEvent protocol
    public let name: String = "EmptyEvent"

    public var payload: Payload {
        return [:]
    }

    public var registeredTrackers: [String] {
        return []
    }

    //keys
    

    

    
}
```

Please change the code generator accordingly.

ssouzawallace replied:
I think changes in code would be "brute force"
Maybe we should rethink the template in order to do this. I totally agree that is ugly now. 
My intent was just to allow empty payloads.

federicocappelli replied:
Don't worry, this is a minor thing, I will refactor the template later

federicocappelli commented on JustTrack/JEEventsGenerator/main.swift:
```diff
@@ -348,7 +358,7 @@ do {
     let structSwiftFilePath = CommandLine.arguments[2]
     if !FileManager.default.fileExists(atPath: structSwiftFilePath) {
-        throw JEStructsGeneratorError.swiftFileNotFound
+        FileManager.default.createFile(atPath: structSwiftFilePath, contents: .none, attributes: .none)
```
Actually, we want to fail if the .swift file isn't there, let's imagine these scenarios if the file doesn't exist:

1. We silently create and fill the .swift file, that isn't added to the project/target, so the build is completed without error but developer can't use JustTrack and the generated events.
2. We fail the build and display clearly the failures reasons, the developer can amend the path or create the file.

we definitely prefer the 2nd scenario.

ssouzawallace replied:
Since improving the log, IMO, you could put it in this generated file if some exception is raised

federicocappelli replied:
the main issue here is that if the script generates the file that file isn't added to the project, and could be really difficult for the developer to understand what's gone wrong. A build failure is much more direct and clear.

ssouzawallace replied:
When I was using the lib for the first time I had some difficult to figure out what was wrong. I think that errors are essencial to explicit what is going on. Compile time errors are the main reason I'm using JustTrack. But the messages are not clear pointing what is wrong with the .plist

As a developer using this, in my opinion, have a file with the erros inside would give me more confidence that the problem is with my implementation and not an internal error. 

I don't want to be annoying. Just clarifying what I was thinking since the raised errors messages be improved.

federicocappelli replied:
Absolutely! @ssouzawallace thank you very much :) your help is much appreciated.
I've raised 2 issues, have a look: [Issues](https://github.com/justeat/JustTrack/issues) and we can use this PR or implement it separately.

About the errors: I think that the best approach is to remove the crashes and improve the error logging at compiling time, making the script fail and printing the errors in the .swift file.

federicocappelli reviewed (changes requested):
I've added some comments, but these are a lot of changes in order to support a corner case.

I'm not convinced that we should support empty Events AT ALL, this is prone to error, especially if the .plist file contains hundreds of events and is maintained and generated by non-developers. The build failure helps to debug the .plist file itself, with your implementation every failure is silent and is impossible to understand what's going wrong.

Another option is to log all the issues, like empty event, missing file etc but proper logging with the swift script is difficult and displayed only in the XCode "Report navigator", generally inspected only in case of build failures.

Good points:

- The swift generator should generate events without payload (but not without trackers IMO)
- for `index in 0...keys.count-1` need to be removed

alannichols reviewed (commented):
Agree with Fede's points.  Definitely worth discussing these changes further.

ssouzawallace committed:
Fixing some requested changes

- Removing spaces in enums attributions
- Inserting guards to check .plist info
- Rasing more exceptions

ssouzawallace committed:
Fixing whitespaces

ssouzawallace committed:
Update main.swift

ssouzawallace commented:
IDK if Github notifies when requested changes are made.
@federicocappelli just doing a heads up. I you already saw this PR after the changes I apologize.

federicocappelli commented:
yes @ssouzawallace i'll review asap

federicocappelli commented:
@ssouzawallace i've converted your PR in a series of feature requests and enhancements request. Have a look: https://github.com/justeat/JustTrack/issues
I can't merge it right now, Some other changes has been planned in parallel so I'll implement everything for 1.0.1 using part of your code.

Thanks for the help ;)

federicocappelli commented:
1.0.1 released with your suggestions. thanks
